posthog/hogql/printer.py

```diff
@@ -439,30 +441,22 @@ def visit_call(self, node: ast.Call):
                         f"Aggregation '{node.name}' cannot be nested inside another aggregation '{stack_node.name}'."
                     )
 
-            translated_args = ", ".join([self.visit(arg) for arg in node.args])
-            if node.distinct:
-                translated_args = f"DISTINCT {translated_args}"
-            return f"{node.name}({translated_args})"
+            args = [self.visit(arg) for arg in node.args]
+            params = [self.visit(param) for param in node.params] if node.params is not None else None
 
-        elif node.name in CLICKHOUSE_FUNCTIONS:
-            clickhouse_name, min_args, max_args = CLICKHOUSE_FUNCTIONS[node.name]
+            params_part = f"({', '.join(params)})" if params is not None else ""
+            args_part = f"({f'DISTINCT ' if node.distinct else ''}{', '.join(args)})"
+            return f"{func_meta.clickhouse_name}{params_part}{args_part}"
 
-            if min_args is not None and len(node.args) < min_args:
-                if min_args == max_args:
-                    raise HogQLException(
-                        f"Function '{node.name}' expects {min_args} arguments. Passed {len(node.args)}."
-                    )
-                raise HogQLException(
-                    f"Function '{node.name}' expects at least {min_args} arguments. Passed {len(node.args)}."
-                )
+        elif node.name in HOGQL_CLICKHOUSE_FUNCTIONS:
+            func_meta = HOGQL_CLICKHOUSE_FUNCTIONS[node.name]
 
-            if max_args is not None and len(node.args) > max_args:
-                if min_args == max_args:
-                    raise HogQLException(
-                        f"Function '{node.name}' expects {max_args} arguments. Passed {len(node.args)}."
-                    )
-                raise HogQLException(
-                    f"Function '{node.name}' expects at most least {max_args} arguments. Passed {len(node.args)}."
+            validate_function_args(node.args, func_meta.min_args, func_meta.max_args, node.name)
+            if func_meta.min_params:
+                if node.params is None:
+                    raise HogQLException(f"Function '{node.name}' requires parameters in addition to arguments")
+                validate_function_args(
+                    node.params, func_meta.min_params, func_meta.max_params, node.name, argument_term="parameter"
                 )
 
             if self.dialect == "clickhouse":
```

Comment:
Why are aggregate functions in a separate branch from the scalar ones?

Reply:
I don't have a better answer than "legacy hack from the python AST days"

Reply:
Would be nice to consolidate… in another PR